### PR TITLE
refactor(prettier): remove side prettier config [PLAT-1077]

### PR DESCRIPTION
**Pull request by htunnicliff**
## Changes

- Removed custom Prettier config package.
- Removed `eslint-plugin-prettier`.
- ~~Added `eslint-config-prettier` to custom packages for `eslint-config-{jest,cypress,react}`.~~
- Added a `.prettierignore` to the lint-config repo.
- Updated lint-staged order and file extensions.
- **Created a shared Prettier config package, `@side/prettier-config`.**

:point_right: See the new Prettier config package [README](https://github.com/reside-eng/lint-config/tree/htunnicliff/PLAT-1216-refactor-prettier/packages/prettier-config#readme) for more information.

## Critical Info 🚨

When a consumer of any lint package updates, it will need to follow the [setup instructions](https://github.com/reside-eng/lint-config/blob/3551f11937e6ebb331dd86c4cd26bf38f89043a2/packages/prettier-config/README.md) provided within `@side/prettier-config` to ensure formatting continues to work.

## Notes

~~The installation docs for `eslint-config-prettier` ask that the Prettier config always be the very last config extended in a given project's ESLint configuration. There is a risk that projects may consume the lint configs provided in this repo, while adding additional items to their `extends` field that deprioritize the Prettier config.~~

~~This PR includes Prettier as the last extends item for the `eslint-config-{jest,cypress,react}` packages, with the assumption that those packages will be the very last config in the repos that import them.~~

~~An alternative we should consider would be this: Every repo imports `eslint-config-prettier` and adds `prettier` as the last item in the `extends` array. For background, all repos will still need to define a `.prettierrc.json` and `.prettierignore`.~~

:point_right: Thanks to a great insight from @a88zach, Prettier gives us a happy path for what we have been trying to do! Their instructions regarding [configuration sharing](https://prettier.io/docs/en/configuration.html#sharing-configurations) explain how to use a package to share a single Prettier configuration across multiple projects. In response to this insight, this PR adds the `@side/prettier-config` package.

### CODE REVIEW

File: .github/dependabot.yml
Comment (offgriddev) [inline .github/dependabot.yml:6]: Is there a YAML rule that prevents all the unneeded quotations? Scalar data types in yaml are only string and numeric, but there are edge cases, of which this might be one. See: https://stackoverflow.com/questions/19109912/yaml-do-i-need-quotes-for-strings-in-yaml

Would be nice to see some rules enforced on quotes in YAML.

Comment (htunnicliff) [reply .github/dependabot.yml:6]: I think this is an instance of Prettier finding double quotes and replacing them with single quotes. You are correct, though, that if there were _no_ quotes, nothing would have changed. 

If we wanted to, I think we could likely get rid of quotes in a lot of these YAML files.

**Comment by a88zach** [comment]
@htunnicliff, I don't think this alone will take care of all repos since some back-end repos extend the base config directly.  In those cases, this change removes prettier as an extension for that linter.

In addition, and this is probably scope creep, we should publish a shared prettier-config from this library that can be shared with all repos:  https://prettier.io/docs/en/configuration.html#sharing-configurations

**Comment by htunnicliff** [comment]
> @htunnicliff, I don't think this alone will take care of all repos since some back-end repos extend the base config directly. In those cases, this change removes prettier as an extension for that linter.
> 
> In addition, and this is probably scope creep, we should publish a shared prettier-config from this library that can be shared with all repos: [prettier.io/docs/en/configuration.html#sharing-configurations](https://prettier.io/docs/en/configuration.html#sharing-configurations)

@a88zach this is amazing! I had no idea this existed. I think a shared config is exactly what we are looking for here. I will update this PR to incorporate something like that.

**Comment by htunnicliff** [comment]
@hershmire @offgriddev – Based on @a88zach's insights regarding shared config, I updated this PR with Prettier's recommended approach for configuration sharing. The PR description has been updated to reflect the changes.

File: .github/dependabot.yml
Comment (htunnicliff) [reply .github/dependabot.yml:6]: > Would be nice to see some rules enforced on quotes in YAML.

@offgriddev Since we don't lint our YAML files, I don't know if there is anything I can do as part of this PR to add rules for YAML quotes. Are there any other changes you want me to make?

**Comment by htunnicliff** [comment]
Following up here – I think this is ready to merge. One more review and it should be good to go!

**Comment by htunnicliff** [comment]
@hershmire – since many folks (myself included) are about to take time off for the holidays, maybe it would be best to merge this in after we all return. Thoughts?

File: packages/prettier-config/README.md
Comment (hershmire) [inline packages/prettier-config/README.md:26]: Can you make a note that this is only need if for some reason they are not using `@side/...` ESLint config? We should have this included in our ESLint base config so teams don't need to deal with extra config.

File: packages/prettier-config/package.json
Comment (hershmire) [inline packages/prettier-config/package.json:4]: We can make this MIT. It will be a public package and okay to share.

**Review by hershmire — CHANGES_REQUESTED** [review]

File: packages/prettier-config/README.md
Comment (htunnicliff) [reply packages/prettier-config/README.md:26]: It might be worth discussing this a bit further:

Regardless of if we keep Prettier's config within the base or not, all consumer repos will have to do the following no matter what:
- Install `prettier` as a dev dependency.
- Install `@side/prettier-config` as a dev dependency.
- Use a `.prettierrc` file or package.json to point Prettier to use the custom config.
- Add a script to the lint-staged configuration that instructs Prettier to format specified file extensions.
- ~~Add `prettier` as the last config in ESLint's `extends` array.~~

I think that removing the last step (~~strikethroughed~~) makes our Prettier setup harder to grasp, since it is the only step here that would be abstracted away into a different package. 

Aside from our goal to have ESLint contain a single "extends" package, are there other strong reasons why we should keep the Prettier ESLint config out of consumer repos?

Comment (hershmire) [reply packages/prettier-config/README.md:26]: > Install @side/prettier-config as a dev dependency.

I'm questioning why we would need each team to manage the above. Unless they are pulling in another ESLint library that would conflict, our library should be able to take care of this for them. One less thing to setup and manage as a dependency.

File: .github/dependabot.yml
Comment (offgriddev) [reply .github/dependabot.yml:6]: @htunnicliff Sorry I missed this. You got the spirit of it in that the majority of the quotes here are unnecessary. I think we could define an `as-needed` rule for quotes on YAML in prettier, at least from what I gather after quickly looking at their docs. I'll make a card to investigate this, no action required here though.

File: packages/prettier-config/README.md
Comment (htunnicliff) [reply packages/prettier-config/README.md:26]: Hey @hershmire! – circling back to this after the holidays 🎄 

I think the TL;DR here is that ESLint is great for linting but is subpar at other tasks like formatting, which is why Prettier recommends not using `eslint-plugin-prettier`. 

If our goal is to remove configuration steps from individual repos, two options come to mind for doing that:
- Create our own CLI that wraps over Prettier, ESLint, and any other libraries you want to abstract.
- Use a monorepo (which would mean there is only one set of configuration).
 
Since both of those would be pretty costly, I think that adopting the `@side/prettier-config` approach will be our best option for the time being.

Comment (hershmire) [reply packages/prettier-config/README.md:26]: Yes, apologies. I don't have an issue with `@side/prettier-config`. For some reason I thought that was referring to the `eslint-config-prettier` which [turns off all rules that are unnecessary or might conflict with Prettier](https://github.com/prettier/eslint-config-prettier). This is what I'm saying we should include in our library so each team doesn't need to manage.

Comment (htunnicliff) [reply packages/prettier-config/README.md:26]: Gotcha! 

I think the only tricky thing with abstracting `eslint-config-prettier` into ESLint configs (as opposed to requiring it in each consumer repo) is that repos will become susceptible to having rules that conflict with Prettier. Here's how that would occur:

```jsonc
// .eslintrc
{
  "extends": [
    "@side/eslint-config-react",
    "some-one-off-config-for-a-plugin" // adds lint rules that Prettier breaks
  ]
}
```

This scenario will be prevented if `eslint-config-prettier` is included at the consumer level, which is instructed in the `@side/prettier-config` setup.

```jsonc
// .eslintrc
{
  "extends": [
    "@side/eslint-config-react",
    "some-one-off-config-for-a-plugin", // adds lint rules that Prettier breaks
   "prettier" // ensures Prettier rules take precedence
  ]
}
```

Comment (htunnicliff) [reply packages/prettier-config/README.md:26]: @hershmire do you have an idea of how you would prevent this scenario if `eslint-config-prettier` was abstracted? As 

far as I can tell, the most reliable way to prevent this buggy scenario would be having consumer repos include `eslint-config-prettier` as the last config in `extends`, which is depicted in the second code block from the previous comment.

Comment (hershmire) [reply packages/prettier-config/README.md:26]: Yeah, the scenario you're suggesting could come up but I would think that's very rare circumstances. I'd imagine it could come up more often if a team was to pull in another library in general. But even then, that might be a very rare scenario. And the team could just turn off those offending rules. What I'm after is minimizing the extra boilerplate for teams. 

I'm okay with moving forward as is for now since I see us using `eslint-config-prettier` with our (not yet created) lint config libs that are specifically tailored for our apps – i.e. `@side/eslint-config-next` for UI apps. At this point, my expectation for UI apps would be to just have this in their .eslintrc.js file:

```js
module.exports = {
  extends: '@side/eslint-config-next',
}
```

**Review by hershmire — APPROVED** [review]
Please get a review from @prescottprue so he's aware of this change and can plan out the changes that will be needed for all our apps and libs.

Thanks for all your work on this!

File: packages/prettier-config/package.json
Comment (prescottprue) [inline packages/prettier-config/package.json:13]: Shouldn't we have prettier as a peer dependency of this package like we use to with the base?

File: .prettierrc.js
Comment (prescottprue) [inline .prettierrc.js:3]: Not trying to be too meta, but one we release the new package we could use `@side/prettier-config` here as well (or remove it and configure within the package.json like you callout in the README)

File: package.json
Comment (prescottprue) [inline package.json:26]: 👏

File: packages/prettier-config/index.json
Comment (prescottprue) [inline packages/prettier-config/index.json:4]: 👏

**Comment by prescottprue** [comment]
Really nice work on this @htunnicliff, thanks for all of the detailed explanation. One thought I had: It seems like this is a breaking change for consumers since they will need to change configurations and install a new package in order to preserve formatting from prettier -or am I misunderstanding? If so we should make sure to mark it as such in the PR title when merging or else teams may inadvertently update without making the necessary changes

File: packages/prettier-config/README.md
Comment (htunnicliff) [reply packages/prettier-config/README.md:26]: Sounds good! I will add `eslint-config-prettier` back in to the other lint packages so that consumer repos do not need to add it. I will also update the README for the prettier-config package.

File: .prettierrc.js
Comment (htunnicliff) [reply .prettierrc.js:3]: I'm not sure about the circular nature of the dependencies there, but I think that is a cool idea! Would it be okay to explore that in a follow up PR?

**Comment by htunnicliff** [comment]
> Really nice work on this @htunnicliff, thanks for all of the detailed explanation. One thought I had: It seems like this is a breaking change for consumers since they will need to change configurations and install a new package in order to preserve formatting from prettier -or am I misunderstanding? If so we should make sure to mark it as such in the PR title when merging or else teams may inadvertently update without making the necessary changes
> 
> Once this is understood I'm ready to approve

@prescottprue you are correct that this PR will cause breaking changes. I included the `BREAKING CHANGE` syntax in the breaking commits, and will be sure that it is represented in the final squashed commit.

**Review by prescottprue — APPROVED** [review]

File: packages/prettier-config/package.json
Comment (hershmire) [reply packages/prettier-config/package.json:13]: We shouldn't if we're not using Prettier for ESLint.

File: .prettierrc.js
Comment (hershmire) [reply .prettierrc.js:3]: I think it's more of a headache to do what you're suggesting Scott. Considering the config isn't that complicated, I'd suggest we stick with just spelling it out in this lib.

**Comment by hershmire** [comment]
Steps to solve the breaking changes can also be included in the body of the commit message so it's automatically added to the release notes.

**Pull request merged**
